chainntnfs/mempool.go

```diff
@@ -0,0 +1,372 @@
+package chainntnfs
+
+import (
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/btcsuite/btcd/btcutil"
+	"github.com/btcsuite/btcd/chaincfg/chainhash"
+	"github.com/btcsuite/btcd/wire"
+	"github.com/lightningnetwork/lnd/lnutils"
+)
+
+// defaultMempoolRefreshInterval defines the default interval to refresh the
+// mempool to check for updates and send notifications to subscribers.
+const defaultMempoolRefreshInterval = 30 * time.Second
+
+// chainConn defines an interface that can be used by the mempool watcher to
+// query transactions in mempool and raw transactions.
+type chainConn interface {
+	// GetRawMempool returns the txids in the current mempool.
+	GetRawMempool() ([]*chainhash.Hash, error)
+
+	// GetRawTransaction returns the raw transaction for the given txid.
+	GetRawTransaction(txHash *chainhash.Hash) (*btcutil.Tx, error)
+}
+
+// TODO(yy): implement this in `btcwallet` instead?
+
+// Mempool defines an internal mempool watcher that's used to watch for the
+// spending of given inputs. This is used either by `BitcoindNotifier` or
+// `BtcdNotifier` depending on the chain backend.
+type Mempool struct {
+	sync.RWMutex
+	wg   sync.WaitGroup
+	quit chan struct{}
+
+	// chain is the chain connection.
+	chain chainConn
+
+	// txids stores the txids in the mempool.
+	//
+	// NOTE: SyncMap is not used here because we expect more writings to
+	// the map.
+	txids map[*chainhash.Hash]struct{}
+
+	// inputs stores the inputs of the txids in the mempool. This is
+	// created for fast lookup.
+	//
+	// NOTE: SyncMap is not used here because we expect more writings to
+	// the map. We only store the outpoints of the inputs plus their
+	// spending txids to save memory.
+	inputs map[*wire.OutPoint]*chainhash.Hash
+
+	// refreshInterval is the interval to refresh the mempool.
+	refreshInterval time.Duration
```

Comment:
Assuming we keep the polling strategy (and don't take advantage of the real-time notifications), then I think we should make this randomized. So we have a min + max, and use a random sample to add some jitter.

Reply:
so the polling strategy is now ditched. For `btcd` we have real-time notifications, for `bitcoind`, we'd still need to make the polling interval random inside `btcwallet`.